include/samurai/load_balancing.hpp

```diff
@@ -0,0 +1,419 @@
+
+            std::vector<mpi::request> req;
+
+            // Actual data exchange **only** with known neighbours of the mesh
+            const auto& neighbours = mesh.mpi_neighbourhood();
+
+            // Phase 2: non-blocking cell sends
+            samurai::times::timers.start("load_balancing_send_cells");
+            // Non-blocking send to each neighbour (possibly empty message)
+            for (const auto& nbr : neighbours)
+            {
+                int rank = nbr.rank;
+                if (rank == world.rank())
+                {
+                    continue;
+                }
+
+                CellArray_t to_send = {payload[static_cast<size_t>(rank)], false};
+                req.push_back(world.isend(rank, 17, to_send));
+            }
+            samurai::times::timers.stop("load_balancing_send_cells");
+
+            // Phase 3: cell reception
+            samurai::times::timers.start("load_balancing_recv_cells");
+            // Blocking reception from each neighbour
+            for (const auto& nbr : neighbours)
+            {
+                int rank = nbr.rank;
+                if (rank == world.rank())
+                {
+                    continue;
+                }
+
+                CellArray_t to_rcv;
+                world.recv(rank, 17, to_rcv);
+
+                samurai::for_each_interval(to_rcv,
+                                           [&](std::size_t level, const auto& interval, const auto& index)
+                                           {
+                                               new_cl[level][index].add_interval(interval);
+                                           });
+            }
+            samurai::times::timers.stop("load_balancing_recv_cells");
+
+            samurai::times::timers.start("load_balancing_wait");
+            boost::mpi::wait_all(req.begin(), req.end());
+            samurai::times::timers.stop("load_balancing_wait");
+
+            samurai::times::timers.start("load_balancing_construct_mesh");
+            Mesh_t new_mesh(new_cl, mesh);
+            samurai::times::timers.stop("load_balancing_construct_mesh");
+
+            samurai::times::timers.stop("load_balancing_mesh_update");
+
+            return new_mesh;
+        }
+
+        template <class Mesh_t, class Weight_t, class Field_t, class... Fields>
+        void load_balance(Mesh_t& mesh, Weight_t& weight, Field_t& field, Fields&... kw)
+        {
+            // Early check: no load balancing with single process
+            boost::mpi::communicator world;
+            if (world.size() <= 1)
+            {
+                std::cout << "Process " << world.rank() << " : Single MPI process detected, load balancing ignored" << std::endl;
+                return;
+            }
+
+            samurai::times::timers.start("load_balancing");
+
+            // Compute flags for this single pass
+            auto flags = static_cast<Flavor&>(*this).load_balance_impl(mesh, weight);
```

Comment:
It could be better to see the CRTP usage more clearly here. We will do that in a next PR.